File: backend/app/utils/error_handler.py
"""
错误处理工具模块
提供错误分类、熔断机制、降级策略和监控功能
"""

import logging
import time
import traceback
from enum import Enum
from typing import Dict, Any, Optional, Callable, List
from functools import wraps
from datetime import datetime, timedelta
import json

logger = logging.getLogger(__name__)


class ErrorSeverity(Enum):
    """错误严重程度"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ErrorCategory(Enum):
    """错误分类"""
    NETWORK = "network"
    DATABASE = "database"
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    VALIDATION = "validation"
    EXTERNAL_SERVICE = "external_service"
    INTERNAL = "internal"
    TIMEOUT = "timeout"
    RATE_LIMIT = "rate_limit"


class CircuitBreakerState(Enum):
    """熔断器状态"""
    CLOSED = "closed"      # 正常状态
    OPEN = "open"          # 熔断状态
    HALF_OPEN = "half_open"  # 半开状态


class ErrorInfo:
    """错误信息封装"""
    
    def __init__(
        self,
        error: Exception,
        category: ErrorCategory,
        severity: ErrorSeverity = ErrorSeverity.MEDIUM,
        context: Optional[Dict[str, Any]] = None,
        user_message: Optional[str] = None
    ):
        self.error = error
        self.category = category
        self.severity = severity
        self.context = context or {}
        self.user_message = user_message
        self.timestamp = datetime.now()
        self.traceback = traceback.format_exc()
        
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典格式"""
        return {
            "error_type": type(self.error).__name__,
            "error_message": str(self.error),
            "category": self.category.value,
            "severity": self.severity.value,
            "context": self.context,
            "user_message": self.user_message,
            "timestamp": self.timestamp.isoformat(),
            "traceback": self.traceback
        }


class CircuitBreaker:
    """熔断器实现"""
    
    def __init__(
        self,
        failure_threshold: int = 5,
        recovery_timeout: int = 60,
        expected_exception: type = Exception,
        half_open_max_calls: int = 3
    ):
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.expected_exception = expected_exception
        self.half_open_max_calls = half_open_max_calls
        
        self.failure_count = 0
        self.last_failure_time = None
        self.state = CircuitBreakerState.CLOSED
        self.half_open_calls = 0
        
    def __call__(self, func: Callable) -> Callable:
        """装饰器实现"""
        @wraps(func)
        def wrapper(*args, **kwargs):
            if self.state == CircuitBreakerState.OPEN:
                if self._should_attempt_reset():
                    self.state = CircuitBreakerState.HALF_OPEN
                    self.half_open_calls = 0
                else:
                    raise Exception("Service unavailable - circuit breaker is OPEN")
            
            try:
                result = func(*args, **kwargs)
                
                if self.state == CircuitBreakerState.HALF_OPEN:
                    self.half_open_calls += 1
                    if self.half_open_calls >= self.half_open_max_calls:
                        self._reset()
                
                return result
                
            except self.expected_exception as e:
                self._record_failure()
                raise e
                
        return wrapper
    
    def _should_attempt_reset(self) -> bool:
        """检查是否应该尝试重置熔断器"""
        return (
            self.last_failure_time and
            time.time() - self.last_failure_time >= self.recovery_timeout
        )
    
    def _record_failure(self):
        """记录失败"""
        self.failure_count += 1
        self.last_failure_time = time.time()
        
        if self.failure_count >= self.failure_threshold:
            self.state = CircuitBreakerState.OPEN
            logger.warning(f"Circuit breaker OPENED for {func.__name__}")
    
    def _reset(self):
        """重置熔断器"""
        self.failure_count = 0
        self.last_failure_time = None
        self.state = CircuitBreakerState.CLOSED
        self.half_open_calls = 0
        logger.info("Circuit breaker RESET to CLOSED")


class RetryPolicy:
    """重试策略"""
    
    def __init__(
        self,
        max_attempts: int = 3,
        base_delay: float = 1.0,
        max_delay: float = 60.0,
        exponential_base: float = 2.0,
        jitter: bool = True
    ):
        self.max_attempts = max_attempts
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.exponential_base = exponential_base
        self.jitter = jitter
    
    def __call__(self, func: Callable) -> Callable:
        """装饰器实现"""
        @wraps(func)
        def wrapper(*args, **kwargs):
            last_exception = None
            
            for attempt in range(self.max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    
                    if attempt == self.max_attempts - 1:
                        break
                    
                    delay = self._calculate_delay(attempt)
                    logger.warning(f"Attempt {attempt + 1} failed for {func.__name__}: {e}. Retrying in {delay:.2f}s")
                    time.sleep(delay)
            
            raise last_exception
            
        return wrapper
    
    def _calculate_delay(self, attempt: int) -> float:
        """计算重试延迟"""
        delay = self.base_delay * (self.exponential_base ** attempt)
        delay = min(delay, self.max_delay)
        
        if self.jitter:
            import random
            delay *= (0.5 + random.random() * 0.5)
        
        return delay


class FallbackHandler:
    """降级处理器"""
    
    def __init__(self):
        self.fallbacks: Dict[str, Callable] = {}
    
    def register(self, service_name: str, fallback_func: Callable):
        """注册降级函数"""
        self.fallbacks[service_name] = fallback_func
        logger.info(f"Registered fallback for {service_name}")
    
    def get_fallback(self, service_name: str) -> Optional[Callable]:
        """获取降级函数"""
        return self.fallbacks.get(service_name)
    
    def execute_fallback(self, service_name: str, *args, **kwargs):
        """执行降级函数"""
        fallback = self.get_fallback(service_name)
        if fallback:
            logger.info(f"Executing fallback for {service_name}")
            return fallback(*args, **kwargs)
        else:
            raise Exception(f"No fallback registered for {service_name}")


class ErrorMonitor:
    """错误监控器"""
    
    def __init__(self, max_errors: int = 1000):
        self.errors: List[ErrorInfo] = []
        self.max_errors = max_errors
        self.error_counts: Dict[str, int] = {}
        self.error_counts_by_category: Dict[ErrorCategory, int] = {}
        self.error_counts_by_severity: Dict[ErrorSeverity, int] = {}
    
    def record_error(self, error_info: ErrorInfo):
        """记录错误"""
        self.errors.append(error_info)
        
        # 保持错误列表大小
        if len(self.errors) > self.max_errors:
            self.errors.pop(0)
        
        # 更新计数
        error_key = f"{type(error_info.error).__name__}:{error_info.category.value}"
        self.error_counts[error_key] = self.error_counts.get(error_key, 0) + 1
        self.error_counts_by_category[error_info.category] = \
            self.error_counts_by_category.get(error_info.category, 0) + 1
        self.error_counts_by_severity[error_info.severity] = \
            self.error_counts_by_severity.get(error_info.severity, 0) + 1
        
        # 记录日志
        log_level = {
            ErrorSeverity.LOW: logging.INFO,
            ErrorSeverity.MEDIUM: logging.WARNING,
            ErrorSeverity.HIGH: logging.ERROR,
            ErrorSeverity.CRITICAL: logging.CRITICAL
        }.get(error_info.severity, logging.WARNING)
        
        logger.log(
            log_level,
            f"Error recorded: {error_info.to_dict()}"
        )
    
    def get_error_summary(self) -> Dict[str, Any]:
        """获取错误摘要"""
        recent_errors = [
            error for error in self.errors
            if error.timestamp > datetime.now() - timedelta(hours=24)
        ]
        
        return {
            "total_errors": len(self.errors),
            "recent_errors_24h": len(recent_errors),
            "error_counts": self.error_counts,
            "errors_by_category": {
                category.value: count
                for category, count in self.error_counts_by_category.items()
            },
            "errors_by_severity": {
                severity.value: count
                for severity, count in self.error_counts_by_severity.items()
            },
            "recent_critical_errors": [
                error.to_dict() for error in recent_errors
                if error.severity == ErrorSeverity.CRITICAL
            ]
        }


# 全局实例
error_monitor = ErrorMonitor()
fallback_handler = FallbackHandler()


def classify_error(error: Exception) -> tuple[ErrorCategory, ErrorSeverity]:
    """错误分类函数"""
    error_type = type(error).__name__
    error_message = str(error).lower()
    
    # 网络错误
    if any(keyword in error_message for keyword in [
        "connection", "timeout", "network", "dns", "socket"
    ]) or error_type in [
        "ConnectionError", "TimeoutError", "NetworkError"
    ]:
        return ErrorCategory.NETWORK, ErrorSeverity.HIGH
    
    # 数据库错误
    if any(keyword in error_message for keyword in [
        "database", "sql", "connection", "constraint", "duplicate"
    ]) or error_type in [
        "DatabaseError", "IntegrityError", "OperationalError"
    ]:
        return ErrorCategory.DATABASE, ErrorSeverity.HIGH
    
    # 认证错误
    if any(keyword in error_message for keyword in [
        "unauthorized", "authentication", "login", "credential"
    ]) or error_type in [
        "AuthenticationError", "UnauthorizedError"
    ]:
        return ErrorCategory.AUTHENTICATION, ErrorSeverity.MEDIUM
    
    # 授权错误
    if any(keyword in error_message for keyword in [
        "forbidden", "permission", "access denied", "unauthorized"
    ]) or error_type in [
        "PermissionError", "ForbiddenError"
    ]:
        return ErrorCategory.AUTHORIZATION, ErrorSeverity.MEDIUM
    
    # 验证错误
    if any(keyword in error_message for keyword in [
        "validation", "invalid", "required", "format"
    ]) or error_type in [
        "ValidationError", "ValueError", "TypeError"
    ]:
        return ErrorCategory.VALIDATION, ErrorSeverity.LOW
    
    # 外部服务错误
    if any(keyword in error_message for keyword in [
        "external", "third party", "api", "service"
    ]) or error_type in [
        "ExternalServiceError", "APIError"
    ]:
        return ErrorCategory.EXTERNAL_SERVICE, ErrorSeverity.HIGH
    
    # 超时错误
    if "timeout" in error_message or error_type in [
        "TimeoutError", "AsyncTimeoutError"
    ]:
        return ErrorCategory.TIMEOUT, ErrorSeverity.HIGH
    
    # 限流错误
    if any(keyword in error_message for keyword in [
        "rate limit", "too many requests", "quota"
    ]) or error_type in [
        "RateLimitError", "TooManyRequestsError"
    ]:
        return ErrorCategory.RATE_LIMIT, ErrorSeverity.MEDIUM
    
    # 默认为内部错误
    return ErrorCategory.INTERNAL, ErrorSeverity.MEDIUM


def get_user_friendly_message(error_info: ErrorInfo) -> str:
    """获取用户友好的错误消息"""
    if error_info.user_message:
        return error_info.user_message
    
    category_messages = {
        ErrorCategory.NETWORK: "网络连接出现问题，请检查您的网络连接后重试",
        ErrorCategory.DATABASE: "数据服务暂时不可用，请稍后重试",
        ErrorCategory.AUTHENTICATION: "身份验证失败，请重新登录",
        ErrorCategory.AUTHORIZATION: "您没有权限执行此操作",
        ErrorCategory.VALIDATION: "输入数据格式不正确，请检查后重试",
        ErrorCategory.EXTERNAL_SERVICE: "外部服务暂时不可用，请稍后重试",
        ErrorCategory.INTERNAL: "系统内部错误，请联系管理员",
        ErrorCategory.TIMEOUT: "请求超时，请稍后重试",
        ErrorCategory.RATE_LIMIT: "请求过于频繁，请稍后再试"
    }
    
    return category_messages.get(error_info.category, "未知错误，请稍后重试")


def handle_error(
    error: Exception,
    context: Optional[Dict[str, Any]] = None,
    user_message: Optional[str] = None
) -> ErrorInfo:
    """统一错误处理函数"""
    category, severity = classify_error(error)
    
    error_info = ErrorInfo(
        error=error,
        category=category,
        severity=severity,
        context=context,
        user_message=user_message
    )
    
    # 记录错误
    error_monitor.record_error(error_info)
    
    return error_info


def with_error_handling(
    fallback_service: Optional[str] = None,
    circuit_breaker: Optional[CircuitBreaker] = None,
    retry_policy: Optional[RetryPolicy] = None
):
    """错误处理装饰器"""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            try:
                # 应用重试策略
                if retry_policy:
                    return retry_policy(func)(*args, **kwargs)
                else:
                    return func(*args, **kwargs)
                    
            except Exception as e:
                error_info = handle_error(
                    e,
                    context={
                        "function": func.__name__,
                        "args": str(args)[:200],  # 限制长度
                        "kwargs": str(kwargs)[:200]
                    }
                )
                
                # 尝试降级处理
                if fallback_service:
                    try:
                        return fallback_handler.execute_fallback(
                            fallback_service, *args, **kwargs
                        )
                    except Exception as fallback_error:
                        logger.error(f"Fallback failed for {fallback_service}: {fallback_error}")
                
                # 重新抛出原始错误
                raise e
                
        # 应用熔断器
        if circuit_breaker:
            wrapper = circuit_breaker(wrapper)
            
        return wrapper
    return decorator

File: frontend/src/utils/errorHandler.ts
/**
 * 前端错误处理工具模块
 * 提供错误分类、用户友好提示、重试机制和错误监控
 */

import React from 'react';

// 错误严重程度
export enum ErrorSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical'
}

// 错误分类
export enum ErrorCategory {
  NETWORK = 'network',
  AUTHENTICATION = 'authentication',
  AUTHORIZATION = 'authorization',
  VALIDATION = 'validation',
  EXTERNAL_SERVICE = 'external_service',
  INTERNAL = 'internal',
  TIMEOUT = 'timeout',
  RATE_LIMIT = 'rate_limit'
}

// 错误信息接口
export interface ErrorInfo {
  error: Error | AxiosError;
  category: ErrorCategory;
  severity: ErrorSeverity;
  context?: Record<string, any>;
  userMessage?: string;
  timestamp: Date;
  stack?: string;
}

// 错误监控器
export class ErrorMonitor {
  private errors: ErrorInfo[] = [];
  private maxErrors: number = 1000;
  private errorCounts: Record<string, number> = {};
  private errorCountsByCategory: Record<ErrorCategory, number> = {
    [ErrorCategory.NETWORK]: 0,
    [ErrorCategory.AUTHENTICATION]: 0,
    [ErrorCategory.AUTHORIZATION]: 0,
    [ErrorCategory.VALIDATION]: 0,
    [ErrorCategory.EXTERNAL_SERVICE]: 0,
    [ErrorCategory.INTERNAL]: 0,
    [ErrorCategory.TIMEOUT]: 0,
    [ErrorCategory.RATE_LIMIT]: 0
  };
  private errorCountsBySeverity: Record<ErrorSeverity, number> = {
    [ErrorSeverity.LOW]: 0,
    [ErrorSeverity.MEDIUM]: 0,
    [ErrorSeverity.HIGH]: 0,
    [ErrorSeverity.CRITICAL]: 0
  };

  constructor(maxErrors: number = 1000) {
    this.maxErrors = maxErrors;
  }

  recordError(errorInfo: ErrorInfo): void {
    this.errors.push(errorInfo);
    
    // 保持错误列表大小
    if (this.errors.length > this.maxErrors) {
      this.errors.shift();
    }
    
    // 更新计数
    const errorKey = `${errorInfo.error.constructor.name}:${errorInfo.category}`;
    this.errorCounts[errorKey] = (this.errorCounts[errorKey] || 0) + 1;
    this.errorCountsByCategory[errorInfo.category] = 
      (this.errorCountsByCategory[errorInfo.category] || 0) + 1;
    this.errorCountsBySeverity[errorInfo.severity] = 
      (this.errorCountsBySeverity[errorInfo.severity] || 0) + 1;
    
    // 记录到控制台
    const logLevel = {
      [ErrorSeverity.LOW]: 'info',
      [ErrorSeverity.MEDIUM]: 'warn',
      [ErrorSeverity.HIGH]: 'error',
      [ErrorSeverity.CRITICAL]: 'error'
    }[errorInfo.severity] || 'warn';
    
    (console as any)[logLevel]('Error recorded:', errorInfo);
  }

  getErrorSummary(): Record<string, any> {
    const now = new Date();
    const oneDayAgo = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    
    const recentErrors = this.errors.filter(
      error => error.timestamp > oneDayAgo
    );
    
    return {
      totalErrors: this.errors.length,
      recentErrors24h: recentErrors.length,
      errorCounts: this.errorCounts,
      errorsByCategory: Object.fromEntries(
        Object.entries(this.errorCountsByCategory).map(([key, value]) => [key, value])
      ),
      errorsBySeverity: Object.fromEntries(
        Object.entries(this.errorCountsBySeverity).map(([key, value]) => [key, value])
      ),
      recentCriticalErrors: recentErrors
        .filter(error => error.severity === ErrorSeverity.CRITICAL)
        .map(error => this.errorInfoToObject(error))
    };
  }

  private errorInfoToObject(errorInfo: ErrorInfo): Record<string, any> {
    return {
      errorType: errorInfo.error.constructor.name,
      errorMessage: errorInfo.error.message,
      category: errorInfo.category,
      severity: errorInfo.severity,
      context: errorInfo.context,
      userMessage: errorInfo.userMessage,
      timestamp: errorInfo.timestamp.toISOString(),
      stack: errorInfo.stack
    };
  }
}

// 重试策略
export class RetryPolicy {
  private maxAttempts: number;
  private baseDelay: number;
  private maxDelay: number;
  private exponentialBase: number;
  private jitter: boolean;

  constructor(options: {
    maxAttempts?: number;
    baseDelay?: number;
    maxDelay?: number;
    exponentialBase?: number;
    jitter?: boolean;
  } = {}) {
    this.maxAttempts = options.maxAttempts || 3;
    this.baseDelay = options.baseDelay || 1000;
    this.maxDelay = options.maxDelay || 60000;
    this.exponentialBase = options.exponentialBase || 2;
    this.jitter = options.jitter !== false;
  }

  async execute<T>(
    fn: () => Promise<T>,
    context?: Record<string, any>
  ): Promise<T> {
    let lastError: Error;
    
    for (let attempt = 0; attempt < this.maxAttempts; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error as Error;
        
        if (attempt === this.maxAttempts - 1) {
          break;
        }
        
        const delay = this.calculateDelay(attempt);
        console.warn(`Attempt ${attempt + 1} failed. Retrying in ${delay}ms:`, error);
        
        await this.sleep(delay);
      }
    }
    
    throw lastError!;
  }

  private calculateDelay(attempt: number): number {
    let delay = this.baseDelay * Math.pow(this.exponentialBase, attempt);
    delay = Math.min(delay, this.maxDelay);
    
    if (this.jitter) {
      delay *= (0.5 + Math.random() * 0.5);
    }
    
    return delay;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// 降级处理器
export class FallbackHandler {
  private fallbacks: Record<string, (...args: any[]) => any> = {};

  register(serviceName: string, fallbackFunc: (...args: any[]) => any): void {
    this.fallbacks[serviceName] = fallbackFunc;
    console.info(`Registered fallback for ${serviceName}`);
  }

  getFallback(serviceName: string): ((...args: any[]) => any) | undefined {
    return this.fallbacks[serviceName];
  }

  async executeFallback(serviceName: string, ...args: any[]): Promise<any> {
    const fallback = this.getFallback(serviceName);
    if (fallback) {
      console.info(`Executing fallback for ${serviceName}`);
      return await fallback(...args);
    } else {
      throw new Error(`No fallback registered for ${serviceName}`);
    }
  }
}

// Axios错误类型定义
interface AxiosError extends Error {
  config?: any;
  code?: string;
  request?: any;
  response?: {
    data?: any;
    status?: number;
    headers?: any;
  };
  isAxiosError?: boolean;
}

// 错误分类函数
export function classifyError(error: Error | AxiosError): [ErrorCategory, ErrorSeverity] {
  const errorType = error.constructor.name;
  const errorMessage = error.message.toLowerCase();
  
  // 网络错误
  if (
    errorMessage.includes('network') ||
    errorMessage.includes('connection') ||
    errorMessage.includes('timeout') ||
    errorMessage.includes('fetch') ||
    errorType === 'NetworkError' ||
    errorType === 'TypeError' && errorMessage.includes('failed to fetch')
  ) {
    return [ErrorCategory.NETWORK, ErrorSeverity.HIGH];
  }
  
  // Axios网络错误
  if ((error as AxiosError).isAxiosError && !(error as AxiosError).response) {
    return [ErrorCategory.NETWORK, ErrorSeverity.HIGH];
  }
  
  // 认证错误
  if (
    errorMessage.includes('unauthorized') ||
    errorMessage.includes('authentication') ||
    errorMessage.includes('login') ||
    errorMessage.includes('credential') ||
    (error as AxiosError).response?.status === 401
  ) {
    return [ErrorCategory.AUTHENTICATION, ErrorSeverity.MEDIUM];
  }
  
  // 授权错误
  if (
    errorMessage.includes('forbidden') ||
    errorMessage.includes('permission') ||
    errorMessage.includes('access denied') ||
    (error as AxiosError).response?.status === 403
  ) {
    return [ErrorCategory.AUTHORIZATION, ErrorSeverity.MEDIUM];
  }
  
  // 验证错误
  if (
    errorMessage.includes('validation') ||
    errorMessage.includes('invalid') ||
    errorMessage.includes('required') ||
    errorMessage.includes('format') ||
    (error as AxiosError).response?.status === 400
  ) {
    return [ErrorCategory.VALIDATION, ErrorSeverity.LOW];
  }
  
  // 外部服务错误
  if (
    errorMessage.includes('external') ||
    errorMessage.includes('third party') ||
    errorMessage.includes('api') ||
    errorMessage.includes('service')
  ) {
    return [ErrorCategory.EXTERNAL_SERVICE, ErrorSeverity.HIGH];
  }
  
  // 服务器错误
  const axiosError = error as AxiosError;
  if (axiosError.response && typeof axiosError.response.status === 'number' && axiosError.response.status >= 500) {
    return [ErrorCategory.EXTERNAL_SERVICE, ErrorSeverity.HIGH];
  }
  
  // 超时错误
  if (
    errorMessage.includes('timeout') ||
    errorType === 'TimeoutError' ||
    (error as AxiosError).code === 'ECONNABORTED'
  ) {
    return [ErrorCategory.TIMEOUT, ErrorSeverity.HIGH];
  }
  
  // 限流错误
  if (
    errorMessage.includes('rate limit') ||
    errorMessage.includes('too many requests') ||
    errorMessage.includes('quota') ||
    (error as AxiosError).response?.status === 429
  ) {
    return [ErrorCategory.RATE_LIMIT, ErrorSeverity.MEDIUM];
  }
  
  // 默认为内部错误
  return [ErrorCategory.INTERNAL, ErrorSeverity.MEDIUM];
}

// 获取用户友好的错误消息
export function getUserFriendlyMessage(errorInfo: ErrorInfo): string {
  if (errorInfo.userMessage) {
    return errorInfo.userMessage;
  }
  
  const categoryMessages: Record<ErrorCategory, string> = {
    [ErrorCategory.NETWORK]: '网络连接出现问题，请检查您的网络连接后重试',
    [ErrorCategory.AUTHENTICATION]: '身份验证失败，请重新登录',
    [ErrorCategory.AUTHORIZATION]: '您没有权限执行此操作',
    [ErrorCategory.VALIDATION]: '输入数据格式不正确，请检查后重试',
    [ErrorCategory.EXTERNAL_SERVICE]: '外部服务暂时不可用，请稍后重试',
    [ErrorCategory.INTERNAL]: '系统内部错误，请联系管理员',
    [ErrorCategory.TIMEOUT]: '请求超时，请稍后重试',
    [ErrorCategory.RATE_LIMIT]: '请求过于频繁，请稍后再试'
  };
  
  return categoryMessages[errorInfo.category] || '未知错误，请稍后重试';
}

// 统一错误处理函数
export function handleError(
  error: Error | AxiosError,
  context?: Record<string, any>,
  userMessage?: string
): ErrorInfo {
  const [category, severity] = classifyError(error);
  
  const errorInfo: ErrorInfo = {
    error,
    category,
    severity,
    context,
    userMessage,
    timestamp: new Date(),
    stack: error.stack
  };
  
  // 记录错误
  errorMonitor.recordError(errorInfo);
  
  return errorInfo;
}

// 错误处理装饰器
export function withErrorHandling(options: {
  fallbackService?: string;
  retryPolicy?: RetryPolicy;
  context?: Record<string, any>;
  userMessage?: string;
} = {}) {
  return function<T extends (...args: any[]) => Promise<any>>(
    target: any,
    propertyName: string,
    descriptor: TypedPropertyDescriptor<T>
  ) {
    const method = descriptor.value!;
    
    descriptor.value = async function(this: any, ...args: any[]) {
      try {
        // 应用重试策略
        if (options.retryPolicy) {
          return await options.retryPolicy.execute(
            () => method.apply(this, args),
            { ...options.context, method: propertyName }
          );
        } else {
          return await method.apply(this, args);
        }
      } catch (error) {
        const errorInfo = handleError(
          error as Error,
          {
            ...options.context,
            method: propertyName,
            args: JSON.stringify(args).substring(0, 200) // 限制长度
          },
          options.userMessage
        );
        
        // 尝试降级处理
        if (options.fallbackService) {
          try {
            return await fallbackHandler.executeFallback(
              options.fallbackService,
              ...args
            );
          } catch (fallbackError) {
            console.error(`Fallback failed for ${options.fallbackService}:`, fallbackError);
          }
        }
        
        // 重新抛出原始错误
        throw error;
      }
    } as T;
    
    return descriptor;
  };
}

// 全局实例
export const errorMonitor = new ErrorMonitor();
export const fallbackHandler = new FallbackHandler();

// 默认重试策略
export const defaultRetryPolicy = new RetryPolicy({
  maxAttempts: 3,
  baseDelay: 1000,
  maxDelay: 10000,
  exponentialBase: 2,
  jitter: true
});

// 错误边界组件的辅助函数
export function createErrorBoundaryFallback(
  error: Error,
  resetErrorBoundary: () => void
): React.ReactElement {
  const errorInfo = handleError(error);
  const userMessage = getUserFriendlyMessage(errorInfo);
  
  return React.createElement('div', {
    className: "error-boundary-fallback p-6 bg-red-50 border border-red-200 rounded-lg"
  }, [
    React.createElement('h2', {
      key: 'title',
      className: "text-xl font-semibold text-red-800 mb-2"
    }, '出现错误'),
    React.createElement('p', {
      key: 'message',
      className: "text-red-600 mb-4"
    }, userMessage),
    React.createElement('button', {
      key: 'retry',
      onClick: resetErrorBoundary,
      className: "px-4 py-2 bg-red-600 text-white rounded hover:bg-red-700 transition-colors"
    }, '重试'),
    ...(process.env.NODE_ENV === 'development' ? [
      React.createElement('details', {
        key: 'details',
        className: "mt-4"
      }, [
        React.createElement('summary', {
          key: 'summary',
          className: "cursor-pointer text-red-500"
        }, '错误详情'),
        React.createElement('pre', {
          key: 'stack',
          className: "mt-2 p-2 bg-red-100 text-xs overflow-auto"
        }, error.stack || 'No stack trace available')
      ])
    ] : [])
  ]);
}

// 导出默认配置
export const ERROR_CONFIG = {
  // 自动重试的错误类型
  RETRYABLE_ERRORS: [
    ErrorCategory.NETWORK,
    ErrorCategory.TIMEOUT,
    ErrorCategory.EXTERNAL_SERVICE
  ],
  
  // 需要用户重新认证的错误
  AUTH_ERRORS: [
    ErrorCategory.AUTHENTICATION
  ],
  
  // 需要降级处理的错误
  FALLBACK_ERRORS: [
    ErrorCategory.EXTERNAL_SERVICE,
    ErrorCategory.NETWORK
  ]
};